File: dotnet/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example4/ChatRoom.cs
namespace MediatorDesignPattern.Example4;

/// <summary>
/// Concrete Mediator - Chat room that coordinates all messaging
/// </summary>
public class ChatRoom : IChatMediator
{
    private readonly List<User> _users = new();
    private readonly string _roomName;

    public ChatRoom(string roomName)
    {
        _roomName = roomName;
        Console.WriteLine($"[CHAT ROOM] '{_roomName}' created.");
    }

    public void RegisterUser(User user)
    {
        _users.Add(user);
        Console.WriteLine($"[CHAT ROOM] {user.Name} joined the room.");
    }

    public void SendMessage(string message, User sender)
    {
        foreach (var user in _users.Where(u => u != sender))
        {
            user.Receive(message, sender);
        }
    }

    public void SendPrivateMessage(string message, User sender, User receiver)
    {
        if (_users.Contains(receiver))
        {
            receiver.Receive($"[Private] {message}", sender);
        }
        else
        {
            Console.WriteLine($"[CHAT ROOM] User {receiver.Name} is not in the room.");
        }
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example4/src/ChatRoom.ts
import { IChatMediator } from './IChatMediator';
import { User } from './User';

export class ChatRoom implements IChatMediator {
  private readonly _users: User[] = [];
  private readonly _roomName: string;

  constructor(roomName: string) {
    this._roomName = roomName;
    console.log(`[CHAT ROOM] '${this._roomName}' created.`);
  }

  registerUser(user: User): void {
    this._users.push(user);
    console.log(`[CHAT ROOM] ${user.name} joined the room.`);
  }

  sendMessage(message: string, sender: User): void {
    for (const user of this._users.filter(u => u !== sender)) {
      user.receive(message, sender);
    }
  }

  sendPrivateMessage(message: string, sender: User, receiver: User): void {
    if (this._users.includes(receiver)) {
      receiver.receive(`[Private] ${message}`, sender);
    } else {
      console.log(`[CHAT ROOM] User ${receiver.name} is not in the room.`);
    }
  }
}
